lent parameter: invalid C codegen

While trying to work around #16897 I got the compiler to generate invalid C with lent input parameters.

Test case

```Nim
type
  Fp[N: static int, T] = object
    big: array[N, T]

type
  QuadraticExt* = concept x
    ## Quadratic Extension concept (like complex)
    type BaseField = auto
    x.c0 is BaseField
    x.c1 is BaseField

{.experimental:"views".}

func prod(r: var QuadraticExt, a, b: lent QuadraticExt) =
  discard

type
  Fp2[N: static int, T] = object
    c0, c1: Fp[N, T]

# This should be passed by reference,
# but concepts do not respect the 24 bytes rule
# or `byref` pragma.
var r, a, b: Fp2[6, uint64]

prod(r, a, b)

```

The proc is defined with pointers:
```Nim
/* section: NIM_merge_PROCS */
N_LIB_PRIVATE N_NIMCALL(void, prod__j3ijsYk5xsFR79akOCPoMqg)(tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* r, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* a, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* b) {
}
```
but is called with values
```Nim
N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
{
	prod__j3ijsYk5xsFR79akOCPoMqg((&r__mSc0P1JKql3o5d2mQ9bx9cMQ), a__IakjUEvgThFFu4dR6kEN0A, b__4KxaakBDXQrcdXChd3FntQ);
}
}
```
Compiler error
```
[...]/nimcache/concept_param/@mconcept_param.nim.c: In function ‘NimMainModule’:
[...]/nimcache/concept_param/@mconcept_param.nim.c:99:64: error: incompatible type for argument 2 of ‘prod__j3ijsYk5xsFR79akOCPoMqg’
   99 |  prod__j3ijsYk5xsFR79akOCPoMqg((&r__mSc0P1JKql3o5d2mQ9bx9cMQ), a__IakjUEvgThFFu4dR6kEN0A, b__4KxaakBDXQrcdXChd3FntQ);
      |                                                                ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                |
      |                                                                tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ
[...]/nimcache/concept_param/@mconcept_param.nim.c:56:143: note: expected ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ *’ but argument is of type ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ’
   56 | ALL(void, prod__j3ijsYk5xsFR79akOCPoMqg)(tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* r, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* a, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* b) {
      |                                                                                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^

[...]/nimcache/concept_param/@mconcept_param.nim.c:99:91: error: incompatible type for argument 3 of ‘prod__j3ijsYk5xsFR79akOCPoMqg’
   99 |  prod__j3ijsYk5xsFR79akOCPoMqg((&r__mSc0P1JKql3o5d2mQ9bx9cMQ), a__IakjUEvgThFFu4dR6kEN0A, b__4KxaakBDXQrcdXChd3FntQ);
      |                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                                           |
      |                                                                                           tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ
[...]/nimcache/concept_param/@mconcept_param.nim.c:56:185: note: expected ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ *’ but argument is of type ‘tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ’
   56 | oMqg)(tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* r, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* a, tyObject_Fp2__SVgH9c6gmNsJfyYVtl0YdSQ* b) {
      |                                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^

Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -O3 -fno-strict-aliasing -fno-ident   -I'[...]/.choosenim/toolchains/nim-#devel/lib' -I[...]/build -o [...]/nimcache/concept_param/@mconcept_param.nim.c.o [...]/nimcache/concept_param/@mconcept_param.nim.c' failed with exit code: 1
```

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -1229,8 +1229,7 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
       if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped):
         if not isMagic(getCurrOwner(c)):
-          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
-                  "' is only allowed in templates and macros or magic procs")
+          localError(c.config, a[^2].info, "'" & typ.sym.name.s & "' is only allowed in templates and macros or magic procs")
 
 
     if hasDefault:

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -1227,12 +1227,9 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
     if hasType:
       typ = semParamType(c, a[^2], constraint)
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind in {skProc, skFunc}:
-        if (typ.kind == tyTyped or typ.kind == tyUntyped) and not isMagic(getCurrOwner(c)):
-          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
+      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped) and not isMagic(getCurrOwner(c)):
+        localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
                   "' is only allowed in templates and macros or magic procs")
-        elif typ.kind == tyLent:
-          localError(c.config, a[^2].info, "`lent` is only allowed in return type")
 
 
     if hasDefault:

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -1227,8 +1227,9 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
     if hasType:
       typ = semParamType(c, a[^2], constraint)
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped) and not isMagic(getCurrOwner(c)):
-        localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
+      if kind in {skProc, skFunc} and (typ.kind == tyTyped or typ.kind == tyUntyped):
+        if not isMagic(getCurrOwner(c)):
+          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
                   "' is only allowed in templates and macros or magic procs")
 
 

--- a/compiler/semtypes.nim
+++ b/compiler/semtypes.nim
@@ -10,7 +10,7 @@
 # this module does the semantic checking of type declarations
 # included from sem.nim
 
-import math
+import std/math
 
 const
   errStringOrIdentNodeExpected = "string or ident node expected"
@@ -1171,6 +1171,7 @@ proc liftParamType(c: PContext, procKind: TSymKind, genericParams: PNode,
   else: discard
 
 proc semParamType(c: PContext, n: PNode, constraint: var PNode): PType =
+  ## Semchecks the type of parameters.
   if n.kind == nkCurlyExpr:
     result = semTypeNode(c, n[0], nil)
     constraint = semNodeKindConstraints(n, c.config, 1)
@@ -1226,9 +1227,13 @@ proc semProcTypeNode(c: PContext, n, genericParams: PNode,
     if hasType:
       typ = semParamType(c, a[^2], constraint)
       # TODO: Disallow typed/untyped in procs in the compiler/stdlib
-      if kind == skProc and (typ.kind == tyTyped or typ.kind == tyUntyped):
-        if not isMagic(getCurrOwner(c)):
-          localError(c.config, a[^2].info, "'" & typ.sym.name.s & "' is only allowed in templates and macros or magic procs")
+      if kind in {skProc, skFunc}:
+        if (typ.kind == tyTyped or typ.kind == tyUntyped) and not isMagic(getCurrOwner(c)):
+          localError(c.config, a[^2].info, "'" & typ.sym.name.s & 
+                  "' is only allowed in templates and macros or magic procs")
+        elif typ.kind == tyLent:
+          localError(c.config, a[^2].info, "`lent` is only allowed in return type")
+
 
     if hasDefault:
       def = a[^1]

--- a/compiler/typeallowed.nim
+++ b/compiler/typeallowed.nim
@@ -57,7 +57,8 @@ proc typeAllowedAux(marker: var IntSet, typ: PType, kind: TSymKind,
   of tyVar, tyLent:
     if kind in {skProc, skFunc, skConst} and (views notin c.features):
       result = t
-    elif t.kind == tyLent and kind != skResult and (views notin c.features):
+    elif t.kind == tyLent and ((kind != skResult and views notin c.features) or
+                              kind == skParam): # lent can't be used as parameters.
       result = t
     else:
       var t2 = skipTypes(t[0], abstractInst-{tyTypeDesc, tySink})

--- a/tests/lent/t16898.nim
+++ b/tests/lent/t16898.nim
@@ -1,5 +1,5 @@
 discard """
-  errormsg: "`lent` is only allowed in return type"
+  errormsg: "invalid type: 'lent QuadraticExt' in this context: 'proc (r: var QuadraticExt, a: lent QuadraticExt, b: lent QuadraticExt){.noSideEffect, gcsafe, locks: 0.}' for proc"
 """
 
 # bug #16898

--- a/tests/lent/t16898.nim
+++ b/tests/lent/t16898.nim
@@ -1,5 +1,5 @@
 discard """
-  errormsg: "lent is only allowed in return type"
+  errormsg: "`lent` is only allowed in return type"
 """
 
 # bug #16898

--- a/tests/lent/t16898.nim
+++ b/tests/lent/t16898.nim
@@ -1,0 +1,31 @@
+discard """
+  errormsg: "lent is only allowed in return type"
+"""
+
+# bug #16898
+type
+  Fp[N: static int, T] = object
+    big: array[N, T]
+
+type
+  QuadraticExt* = concept x
+    ## Quadratic Extension concept (like complex)
+    type BaseField = auto
+    x.c0 is BaseField
+    x.c1 is BaseField
+
+{.experimental:"views".}
+
+func prod(r: var QuadraticExt, a, b: lent QuadraticExt) =
+  discard
+
+type
+  Fp2[N: static int, T] = object
+    c0, c1: Fp[N, T]
+
+# This should be passed by reference,
+# but concepts do not respect the 24 bytes rule
+# or `byref` pragma.
+var r, a, b: Fp2[6, uint64]
+
+prod(r, a, b)

--- a/tests/lent/t17621.nim
+++ b/tests/lent/t17621.nim
@@ -1,5 +1,5 @@
 discard """
-  errormsg: "`lent` is only allowed in return type"
+  errormsg: "invalid type: 'lent Test' in this context: 'proc (self: lent Test)' for proc"
 """
 
 # bug #17621

--- a/tests/lent/t17621.nim
+++ b/tests/lent/t17621.nim
@@ -1,0 +1,15 @@
+discard """
+  errormsg: "lent is only allowed in return type"
+"""
+
+# bug #17621
+{.experimental: "views".}
+
+type Test = ref object
+  foo: int
+
+proc modify(self: lent Test) =
+  self.foo += 1
+
+let test = Test(foo: 12)
+modify(test)

--- a/tests/lent/t17621.nim
+++ b/tests/lent/t17621.nim
@@ -1,5 +1,5 @@
 discard """
-  errormsg: "lent is only allowed in return type"
+  errormsg: "`lent` is only allowed in return type"
 """
 
 # bug #17621

